assembly/src/assembler/mod.rs

```diff
@@ -595,86 +601,148 @@ impl Assembler {
     {
         use ast::Op;
 
-        let mut node_ids: Vec<MastNodeId> = Vec::new();
-        let mut basic_block_builder = BasicBlockBuilder::new(wrapper);
+        let mut body_node_ids: Vec<MastNodeId> = Vec::new();
+        let mut block_builder = BasicBlockBuilder::new(wrapper, mast_forest_builder);
 
         for op in body {
             match op {
                 Op::Inst(inst) => {
-                    if let Some(mast_node_id) = self.compile_instruction(
-                        inst,
-                        &mut basic_block_builder,
-                        proc_ctx,
-                        mast_forest_builder,
-                    )? {
-                        if let Some(basic_block_id) =
-                            basic_block_builder.make_basic_block(mast_forest_builder)?
-                        {
-                            node_ids.push(basic_block_id);
+                    if let Some(node_id) =
+                        self.compile_instruction(inst, &mut block_builder, proc_ctx)?
+                    {
+                        if let Some(basic_block_id) = block_builder.make_basic_block()? {
+                            body_node_ids.push(basic_block_id);
+                        } else if let Some(decorator_ids) = block_builder.drain_decorators() {
+                            block_builder
+                                .mast_forest_builder_mut()
+                                .set_before_enter(node_id, decorator_ids);
                         }
 
-                        node_ids.push(mast_node_id);
+                        body_node_ids.push(node_id);
                     }
                 },
 
                 Op::If { then_blk, else_blk, .. } => {
-                    if let Some(basic_block_id) =
-                        basic_block_builder.make_basic_block(mast_forest_builder)?
-                    {
-                        node_ids.push(basic_block_id);
+                    if let Some(basic_block_id) = block_builder.make_basic_block()? {
+                        body_node_ids.push(basic_block_id);
                     }
 
-                    let then_blk =
-                        self.compile_body(then_blk.iter(), proc_ctx, None, mast_forest_builder)?;
-                    let else_blk =
-                        self.compile_body(else_blk.iter(), proc_ctx, None, mast_forest_builder)?;
+                    let then_blk = self.compile_body(
+                        then_blk.iter(),
+                        proc_ctx,
+                        None,
+                        block_builder.mast_forest_builder_mut(),
+                    )?;
```

Comment:
Not for this PR, but in the future we could move `wrapper` into `ProcedureContext` and then we won't need to pass `None` here and in other similar places.